### PR TITLE
[printing] Deprecate all printing functions accessing the global proof.

### PR DESCRIPTION
**Pull request by ejgallego**
We'd like to handle proofs functionally we thus recommend not to use
printing functions without an explicit context.

We also adapt most of the code, making more explicit where the
printing environment is coming from.

### CODE REVIEW

**Review by herbelin — APPROVED** [review]
I looked at some parts and this looks ok to me. 

If I understand correctly, it is better in some places, e.g. because we had a `env`/`sigma` at hand but did not use it, and not worse in the other cases, because we move some steps upstream the call  to `get_current_context`, reducing at the same time the distance to being purely functional. Is that correct?

Also, if I understand correctly, further PRs are coming to propagate even further the environment and global state functionally. This is great.

**Comment by ejgallego** [comment]
Hi Hugo (@herbelin) thanks a lot for having a look. Indeed, the idea is that we propagate the grab of the env/sigma "upwards", so we reduce the distance to being purely functional. IMHO full migration to a functional style will have to happen in 8.9 but hopefully we can discourage the use of global interfarces in 8.8.

A further patch should deprecate `get_current_context` and friends.

**Comment by maximedenes** [comment]
Will merge as soon as Travis is happy.

**Comment by maximedenes** [comment]
@ejgallego Please indicate when a PR depends on another, until GitHub gets proper support for that. I'm gonna have to redo a sequence of merges to get the history right.

**Comment by ejgallego** [comment]
Travis is fine in my build, indeed this depended on a trivial, now merged PR.

File: interp/constrextern.mli [hunk 1]
```diff
@@ -60,6 +60,19 @@ val set_extern_reference :
 val get_extern_reference :
   unit -> (?loc:Loc.t -> Id.Set.t -> global_reference -> reference)
 
+(** WARNING: The following functions ae evil due to
```
Comment (gares) [inline interp/constrextern.mli]: typo

Comment (ejgallego) [reply interp/constrextern.mli]: typo fixed.

**Comment by ejgallego** [comment]
Travis was OK in the pre-typo tree.

**Comment by maximedenes** [comment]
Green light here: https://travis-ci.org/maximedenes/coq/builds/305396457 so merging to avoid more rebase pain for everyone.

**Comment by herbelin** [comment]
> so merging to avoid more rebase pain for everyone.

May I make this sentence a bit more precise... What would "avoid pain for everyone" is that no branches exist at the same time with merge conflicts, and thus, indeed, reducing the lifespan of a PR is good with this respect.

However, when there are already co-existing branches with merge conflicts, the question is also about finding a good scheduling between branches. I'm thinking in particular to #982, about which I already spent several hours in rebasing, but I'm afraid also about the heavy #930, #991, #268, which, the more we are waiting, the more conflicts are liable to pile up.

Regarding #982, I'm still investigating how to preserve a strict compatibility. I'd like also to write the documentation for the new features. Hopefully, this will be done by the end of this week, with, phew, no more rebase to do!

**Comment by ejgallego** [comment]
@herbelin I am liable for some of these delays, sorry for that. However for #982, a point is that it is really huge in terms of numbers of commits, so it makes things very hard to review. Do you think that some slicing would be possible? I feel like doing more compact PRs helps in merge-time.

On the other hand I am with @ppedrot than in some cases merging is preferrable to reface if `git rerere` is not helping.

That aside, I guess that the period of crazy conflicts is over. Due to the large numbers of pending deprecations pushed at once we have faced an usual conflict rate, but I hope that this is over for the rest of the 8.8 time.

**Comment by ejgallego** [comment]
By the way I guess the reason these PRs you mention are not considered for merge is due to their tags.

So indeed some concrete plan should come up to clear the tags, I ignore what to do. I suffered myself more than 100 rebases with the CAst patch and at some time I thought I would go mad. I guess that's the pain of concurrent development but on the other hand I've personally tried to adapt towards a more compact view in order to avoid large PRs hanging for a while [but even so see #186]

**Pull request merged**